tools/jjb-lint/jjb-lint.sh

```diff
@@ -0,0 +1,376 @@
+		test \
+		-o "$1" \
+		"$2" \
+		2>&1
+}
+
+parse_ansible_output()
+{
+	# This command is a filter that transforms
+	# ansible's output into a more human readable form.
+	#
+	# It echoes to stdout one line per issue,
+	# in the format
+	#   filename: error message
+	#
+	# It returns 1 if any of the ansible jobs failed,
+	# 0 otherwise.
+
+	local status
+	status=0
+
+	while read -r line
+	do
+		# if there is a failure ansible emits a line that includes some json
+		# we use jq to parse the json and output the information we want in the correct format
+		echo "$line" | grep '^failed' | grep -o '{.*}' | jq -r '.item + ": " + .msg' | sed "s|$CWD/||"
+
+		# record if any of the output lines indicates a failure
+		if [[ "$line" == failed:* ]]
+		then
+			status=1
+			if [ "$FAILFAST" = "y" ]
+			then
+				exit 1
+			fi
+		fi
+	done
+
+	return $status
+}
+
+parse_jjb_output()
+{
+	# This command is a filter that transforms
+	# the output of jenkins-jobs messages.
+	#
+	# It echoes to stdout one line per issue,
+	# in the format
+	#   filename: error message
+	#
+	# It returns 1 if jenkins-jobs emitted any
+	# warnings or errors, 0 otherwise.
+
+	local status
+	status=0
+
+	# the first line from run_jjb will be the filename
+	local filename
+	read -r filename
+	filename=${filename##$CWD/}
+
+	while read -r line
+	do
+		# Print the warning line with reduced detail
+		# (remove the python logging prefixes)
+		echo "$filename: $line" | sed -n 's|WARNING:[^ ]*:||p'
+
+		# record if any of the output lines indicates a failure
+		if [[ "$line" == WARNING:* ]]
+		then
+			status=1
+			if [ "$FAILFAST" = "y" ]
+			then
+				exit 1
+			fi
+		fi
+	done
+
+	return $status
+}
+
+ansible_lint()
+{
+	# Usage: ansible_lint FILES...
+	# It saves any output files to $OUTDIR
+	# It logs to ansible.log
+	# Exit status is 1 if any files have issues
+
+	local status
+
+	run_ansible "$OUTDIR" "$@" | pipetrace -a "ansible.log" | parse_ansible_output
+	return $?
+}
+
+jjb_lint()
+{
+	# Usage: jjb_lint FILES...
+	# It saves any output files to $OUTDIR
+	# It logs to jenkins-jobs.log
+	# Exit status is 1 if any files have issues
+	local status
+	status=0
+
+	copy_jjb_config
+
+	for f
+	do
+		# Use uniq before parsing the output because
+		# jenkins-jobs will emit an error for each
+		# job it finds in a YAML file; our yaml files
+		# often the same job repeated for different
+		# environments, so we filter out repeated lines
+		# to avoid noisy output
+
+		run_jjb "$OUTDIR" "$f" | pipetrace -a "jenkins-jobs.log" | uniq | parse_jjb_output
+
+		status=$((status|$?))
+		if [ "$FAILFAST" = "y" ] && [ $status -ne 0 ]
+		then
+			exit 1
+		fi
+	done
+	return $status
+}
+
+copy_jjb_config()
+{
+	# We want to test jenkins-jobs with the
+	# configuration used in production, but
+	# we want to tweak it slightly to avoid
+	# jenkins-jobs contacting jenkins
+
+	run_ansible "jenkins_jobs.ini.out" "${DM_JENKINS_REPO}/playbooks/roles/jenkins/templates/jenkins_jobs.ini.j2" > /dev/null
```

Comment:
Which bits of the config exactly are we tweaking? Presumably the user/password?

Reply:
Ah, that could definitely be better commented.

I will add this comment:

```bash
# If jenkins-jobs finds a configuration file it tries to contact the Jenkins server
# to detect what plugins are installed, unless `query_plugins_info` is set to false
# in the configuration file.
# See <https://docs.openstack.org/infra/jenkins-job-builder/execution.html#jenkins-section>
```